### PR TITLE
Remove claim

### PR DESCRIPTION
**Pull request by JeroenDeDauw**
Note that this is a pull request against the 3.0.x branch, not against master. So we can still make as many 2.x releases as we want before taking this step.

### CODE REVIEW

**Comment by JeroenDeDauw** [comment]
I just had a ton of fun rebasing this and fixing ALL of the conflicts :)

File: RELEASE-NOTES.md [hunk 1]
```diff
@@ -1,7 +1,19 @@
...
+	* The `Claim` class itself has been removed, though `Claim` is now a temporary alias for `Statement`
+	* `Statement` no longer takes a `Claim` in its constructor
+	* `Statement::setClaim` and `Statement::getClaim` have been removed
+	* Removed `ClaimList`
+	* Removed `ClaimListAccess`
+
 ## Version 2.6.0 (dev)
 
+
+## Version 2.5.0 (dev)
```
Comment (thiemowmde) [inline RELEASE-NOTES.md]: Rebase mistake.

File: WikibaseDataModel.php [hunk 2]
```diff
@@ -12,10 +12,12 @@
...
-define( 'WIKIBASE_DATAMODEL_VERSION', '2.6.0 alpha' );
+define( 'WIKIBASE_DATAMODEL_VERSION', '3.0.0 alpha' );
 
 if ( defined( 'MEDIAWIKI' ) ) {
 	call_user_func( function() {
 		require_once __DIR__ . '/WikibaseDataModel.mw.php';
 	} );
 }
+
+class_alias( 'Wikibase\DataModel\Statement\Statement', 'Wikibase\DataModel\Claim\Claim' );
```
Comment (thiemowmde) [inline WikibaseDataModel.php]: Isn't Aliases.php needed in addition to this line?

File: tests/unit/Entity/EntityTest.php [hunk 3]
```diff
@@ -30,13 +30,6 @@
 abstract class EntityTest extends \PHPUnit_Framework_TestCase {
 
 	/**
-	 * Returns several more or less complex claims
-	 *
-	 * @return Claim[]
-	 */
-	public abstract function makeClaims();
```
Comment (thiemowmde) [inline tests/unit/Entity/EntityTest.php]: This seems unrelated. If this is unused (I just checked and yes, it is unused) this should be a separate patch, to be merged before this big one.

**Comment by thiemowmde** [comment]
Which is why I favor an incremental approach (started with #268). I tried to review this (again) and found it almost not possible because of so many things going on the same time. Quite some changes can easily be split into their own patches (as already tried in #268, also see the comments here) and merged before this big one, making review much easier.

File: tests/unit/Statement/StatementTest.php [hunk 4]
```diff
@@ -53,9 +46,9 @@ public function testGetGuid( Statement $statement ) {
 	}
 
 	public function testSetAndGetMainSnak() {
-		$mainSnak = new PropertyNoValueSnak( new PropertyId( 'P42' ) );
-		$statement = new Statement( new Claim( $mainSnak ) );
-		$this->assertSame( $mainSnak, $statement->getMainSnak() );
+		$snak = new PropertyNoValueSnak( new PropertyId( 'P42' ) );
```
Comment (thiemowmde) [inline tests/unit/Statement/StatementTest.php]: Why the rename? Isn't it still a main snak?

File: tests/unit/Statement/StatementListTest.php [hunk 5]
```diff
@@ -265,19 +264,14 @@ public function testCanConstructWithClaimsObjectContainingOnlyStatements() {
 	}
 
 	public function testGivenTraversableWithNonStatements_constructorThrowsException() {
-		$claim = new Claim( new PropertyValueSnak( 42, new StringValue( 'foo' ) ) );
-		$claim->setGuid( 'meh' );
-
-		$claimArray = array(
+		$traversable = new \ArrayObject( array(
```
Comment (thiemowmde) [inline tests/unit/Statement/StatementListTest.php]: I find this very odd now. `ArrayObject` and `stdClass`?

File: tests/unit/Statement/StatementListTest.php [hunk 6]
```diff
@@ -99,7 +99,6 @@ public function testGetBestStatementPerProperty() {
 			$this->getStatement( 3, 'six', Statement::RANK_NORMAL ),
 
 			$this->getStatement( 4, 'seven', Statement::RANK_PREFERRED ),
-			$this->getStatement( 4, 'eight', Claim::RANK_TRUTH ),
```
Comment (thiemowmde) [inline tests/unit/Statement/StatementListTest.php]: Why this change? Seems unrelated.

File: tests/unit/Entity/ItemTest.php [hunk 7]
```diff
@@ -147,9 +90,9 @@ public function itemProvider() {
 		$items[] = $item;
 
 		$item = $item->copy();
-		$item->getStatements()->addNewStatement(
+		$item->getStatements()->addStatement( new Statement(
```
Comment (thiemowmde) [inline tests/unit/Entity/ItemTest.php]: Why this change?

File: tests/unit/Entity/EntityTest.php [hunk 8]
```diff
@@ -425,6 +418,31 @@ public function testHasClaims( Entity $entity ) {
 		$this->assertEquals( count( $entity->getClaims() ) !== 0, $has );
 	}
 
+	/**
+	 * Tests Entity::newClaim
```
Comment (thiemowmde) [inline tests/unit/Entity/EntityTest.php]: Is this a rebase mistake? I think we dropped this broken test some time ago.

File: tests/unit/Claim/ClaimsTest.php [hunk 9]
```diff
@@ -565,12 +568,6 @@ public function provideGetByRank() {
 				Statement::RANK_PREFERRED,
 				new Claims( array( $s2 ) ),
 			),
-			// s2 has RANK_PREFERRED, so doesn't match RANK_TRUTH
-			array(
-				new Claims( array( $s2 ) ),
-				Claim::RANK_TRUTH,
```
Comment (thiemowmde) [inline tests/unit/Claim/ClaimsTest.php]: Quite some changes in this patch set refer to that "truth" thingy, I honestly never understood. Can you please explain or, even better, split this into it's own pull request?

File: tests/unit/Claim/ClaimsTest.php [hunk 10]
```diff
@@ -86,381 +86,384 @@ public function constructorErrorProvider() {
...
+		$this->assertEquals( 4,
+			$statements->indexOf( $statement5 ),
+			'Appending statement by specifying an index out of bounds failed'
+		);
 	}
 
 	public function testIndexOf() {
-		$claims = new Claims();
-		$claimArray = array(
+		$statements = new Claims();
```
Comment (thiemowmde) [inline tests/unit/Claim/ClaimsTest.php]: All the variable name changes here don't make much sense if `$statements` still is of type `Claims` and `$statementArray` still is an array of `makeClaim` calls.

File: src/Statement/Statement.php [hunk 11]
```diff
@@ -180,24 +293,10 @@ public function equals( $target ) {
...
-	}
-
-	/**
-	 * @since 1.0
-	 *
-	 * @return Claim
-	 */
-	public function getClaim() {
-		return $this;
+	private function claimFieldsEqual( Statement $target ) {
```
Comment (thiemowmde) [inline src/Statement/Statement.php]: I know this is just moved code, but it's confusing now. Please inline.

File: src/Statement/Statement.php [hunk 12]
```diff
@@ -30,6 +35,26 @@ class Statement extends Claim {
 	const RANK_NORMAL = 1;
 	const RANK_DEPRECATED = 0;
 
+	/** @deprecated since 2.0 */
+	const RANK_TRUTH = 3;
```
Comment (thiemowmde) [inline src/Statement/Statement.php]: As already suggested in an other comment, I suggest to drop this completely, but do this in a separate pull request that can be merged before this one.

File: tests/unit/Statement/StatementTest.php (hunk 4 above)
Comment (JeroenDeDauw) [inline tests/unit/Statement/StatementTest.php]: Caused by rebase

File: tests/unit/Entity/ItemTest.php (hunk 7 above)
Comment (JeroenDeDauw) [inline tests/unit/Entity/ItemTest.php]: Caused by rebase

File: tests/unit/Statement/StatementListTest.php (hunk 6 above)
Comment (JeroenDeDauw) [inline tests/unit/Statement/StatementListTest.php]: `Claim::RANK_TRUTH` no longer exists

File: tests/unit/Entity/EntityTest.php (hunk 8 above)
Comment (JeroenDeDauw) [inline tests/unit/Entity/EntityTest.php]: Yeah

File: tests/unit/Claim/ClaimsTest.php (hunk 9 above)
Comment (JeroenDeDauw) [inline tests/unit/Claim/ClaimsTest.php]: `Claim::RANK_TRUTH` is a rank claims had, and only claims can have. It has no meaning if you do not have claims

File: src/Statement/Statement.php (hunk 12 above)
Comment (JeroenDeDauw) [inline src/Statement/Statement.php]: It is the intention to drop this completely

**Comment by snaterlicious** [comment]
Why is `Claims` not removed along?

**Comment by JeroenDeDauw** [comment]
Because it is still used in Wikibase.git. Would be nice to also remove it in 3.x, though little reason to try force that into this PR.

**Comment by JeroenDeDauw** [comment]
I'd like to see this merged, so we can make small commits addressing specific things that still need to be done. That's a lot easier than dealing with this clunky change.

**Comment by thiemowmde** [comment]
-1. @JeroenDeDauw, I would love to see this moving forward, but this pull request is so big and does so many things, I find it scary and hard to review. (Also, I will not merge this with the rebase artifacts.) Let's please split this somehow, e.g. rework the `Claims` class before removing `Claim`. I can do this and create separate pull requests, if you want. Please tell me.

**Comment by JeroenDeDauw** [comment]
Go ahead

File: src/Statement/Statement.php [hunk 13]
```diff
@@ -43,12 +68,58 @@ class Statement extends Claim {
...
+	 * @param Snaks|null $qualifiers
 	 * @param ReferenceList|null $references
 	 */
-	public function __construct( Claim $claim, ReferenceList $references = null ) {
-		$this->setClaim( $claim );
-		$this->references = $references ?: new ReferenceList();
+	public function __construct( Snak $mainSnak, Snaks $qualifiers = null, ReferenceList $references = null ) {
+		$this->mainSnak = $mainSnak;
+		$this->qualifiers = $qualifiers === null ? new SnakList() : $qualifiers;
+		$this->references = $references === null ? new ReferenceList() : $references;
```
Comment (thiemowmde) [inline src/Statement/Statement.php]: An other rebase mistake.

**Comment by thiemowmde** [comment]
> Go ahead

Done, see #361, #363 and #364. I will upload more pull request as soon as these are merged.

**Comment by tobijat** [comment]
@thiemowmde please also link the split off changes to the relevant phabricator task https://phabricator.wikimedia.org/T87388

**Comment by thiemowmde** [comment]
* I rebased this change.
* I reverted most of the variable name changes in the `Claims` and `ClaimsTest` classes. They don't help but make the patch pretty much impossible to review.
* I couldn't find any usage of the TRUTH rank and dropped it completely.

Still a -1 because of some odd bits in this big patch. I still want to submit smaller patches to be merged before this.

**Comment by thiemowmde** [comment]
@JeroenDeDauw, I did a final rebase here. Can you please have a look, do a proper review (I may have messed stuff up, please double-check) and +1 if it's ok? I will merge it then.
